finality-aleph/src/justification/requester.rs

```diff
@@ -202,6 +213,67 @@ where
     }
 
     pub fn finalized_number(&self) -> NumberFor<B> {
-        self.client.info().finalized_number
+        self.backend.blockchain().info().finalized_number
+    }
+
+    fn request(&mut self, header: <B as BlockT>::Header) {
+        let number = *header.number();
+        let hash = header.hash();
+        debug!(target: "aleph-justification",
+               "We have block {:?} with hash {:?}. Requesting justification.", number, hash);
+        self.justification_request_scheduler.on_request_sent();
+        self.block_requester.request_justification(&hash, number);
+    }
+
+    // We request justifications for all the children of last finalized block and a justification
+    // for a block of number num on longest branch.
+    // Assuming that we request at the same pace that finalization is progressing, the former ensures
+    // that we are up to date with finalization. On the other hand, the latter enables fast catch up
+    // if we are behind.
+    // We also request the child that it's on the same branch as top_wanted since a fork may happen
+    // somewhere in between them.
+    fn request_targets(&mut self, mut top_wanted: NumberFor<B>) -> Vec<<B as BlockT>::Header> {
+        let blockchain_backend = self.backend.blockchain();
+        let blockchain_info = blockchain_backend.info();
+        let finalized_hash = blockchain_info.finalized_hash;
+
+        let mut targets = blockchain_backend
+            .children(finalized_hash)
+            .unwrap_or_default()
+            .into_iter()
+            .filter_map(|hash| {
+                if let Ok(Some(header)) = blockchain_backend.header(BlockId::Hash(hash)) {
```

Comment:
Hmm, if we have the hash of the block, and we know its number `blockchain_info.finalized_number + 1` then we don't need to query the header, I guess. 

There is even a special type for block_hash + number in aleph finality `BlockHashNum<B>` :) It can be used instead of `Header`

Reply:
Good catch, it simplifies a lot.